src/tasks/worker.ts

```diff
@@ -0,0 +1,276 @@
+import { setTimeout } from 'node:timers/promises';
+import type { AnyObject } from 'pinejs-client-core';
+import { tasks as tasksEnv } from '../config-loader/env';
+import type * as Db from '../database-layer/db';
+import { TransactionClosedError } from '../database-layer/db';
+import * as permissions from '../sbvr-api/permissions';
+import { PinejsClient } from '../sbvr-api/sbvr-utils';
+import { sbvrUtils } from '../server-glue/module';
+import { ajv, channel } from './common';
+import type { PartialTask, TaskHandler, TaskStatus } from './types';
+
+// Map of column names with SBVR names used in SELECT queries
+const selectColumns = Object.entries({
+	id: 'id',
+	'is executed by-handler': 'is_executed_by__handler',
+	'is executed with-parameter set': 'is_executed_with__parameter_set',
+	'is scheduled with-cron expression': 'is_scheduled_with__cron_expression',
+	'attempt count': 'attempt_count',
+	'attempt limit': 'attempt_limit',
+	'is created by-actor': 'is_created_by__actor',
+})
+	.map(([key, value]) => `t."${key}" AS "${value}"`)
+	.join(', ');
+
+// The worker is responsible for executing tasks in the queue. It listens for
+// notifications and polls the database for tasks to execute. It will execute
+// tasks in parallel up to a certain concurrency limit.
+export class Worker {
+	public handlers: Record<string, TaskHandler> = {};
+	private readonly concurrency: number;
+	private readonly interval: number;
+	private executing = 0;
+
+	constructor(private readonly client: PinejsClient) {
+		this.concurrency = tasksEnv.queueConcurrency;
+		this.interval = tasksEnv.queueIntervalMS;
+	}
+
+	// Check if instance can execute more tasks
+	private canExecute(): boolean {
+		return (
+			this.executing < this.concurrency && Object.keys(this.handlers).length > 0
+		);
+	}
+
+	private async execute(task: PartialTask, tx: Db.Tx): Promise<void> {
+		this.executing++;
+		try {
+			// Get specified handler
+			const handler = this.handlers[task.is_executed_by__handler];
+			const startedOnTime = new Date();
+			if (handler == null) {
+				await this.finalize(
+					tx,
+					task,
+					startedOnTime,
+					'failed',
+					'Matching task handler not found',
```

Comment:
I would rather we made this impossible to happen by ensuring code cannot enqueue work without a corresponding handler. That wouldn’t help with outstanding tasks with a deploy that removes handlers in between, but that could also be solved by having a consistency check when initializing the task system at startup to exit immediately if tasks with no handler are found.

My overarching point is we should treat this case as a programmer error, rather than an error we can recover from (by ignoring it, like we do now).

Reply:
We do have a check in the task creation hook to ensure the requested handler exists. I like the idea of also exiting early with an error if there are pending tasks scheduled to be executed with handlers that no longer exist. Although it would be an extra step during init with a possibly costly query (depending on the number of pending tasks we have).

Reply:
Actually, this error should literally never happen. The polling SQL will only select tasks that point to handlers that the task runner is already aware of. I want to keep the check here because I'm paranoid, but I'll make the error message we store in the DB mention that this really should never happen and something is scarily broken. Will also add a comment above this check to mention the same.

Reply:
Sorry, I wasn’t very clear in my comment. What I mean is it should be considered an error to have outstanding jobs in the DB without a matching handler. Otherwise, you risk having orphan work scheduled that code might depend upon it running, leading to consistency/integrity issues. Is that clearer?

Reply:
@dfunckt  I think you were clear before, but my response added some confusion. There are two problems here: (1) the check itself technically not being necessary as the polling SQL should never pick up a task it doesn't know the handler for and (2) a pending task existing in the queue with an unknown handler. I fully responded to (1) and only partially responded to (2). We can add a check for (2) during initialization like you mentioned, but I suspect in the future we may have instances for specific use cases (like we currently have for a couple of things in production) meaning that we may end up with an instance that only knows about the "foobar-handler" for example. But thats just speculation and I think you're idea does make sense in general, so I'll look at adding it along with probably a matching partial index.

Reply:
Yeah, specifically I very much expect that having certain instances that only handle a single type of task is definitely a possibility, for ones that should be isolated either due to their importance (nothing should affect them) or their lack thereof (they shouldn't interact with anything else), and as such not every running pinejs instance would handle every type of task. That said I've realized that an alternative would be to still register them all but tell pinejs to only dequeue a subset for given instances